### PR TITLE
frequency based negative sampling

### PR DESCRIPTION
**Pull request by AdrianKs**
This is a basic frequency based negative sampling approach.

### CODE REVIEW

**Comment by rgemulla** [comment]
What's the state of this PR (it's very useful!)?

**Comment by AdrianKs** [comment]
It needs to be updated to the new structure of the sampler.
I will update it and let you know.

**Comment by AdrianKs** [comment]
The code is ready for review.

File: kge/config-default.yaml [hunk 1]
```diff
@@ -188,7 +188,8 @@ KvsAll:
 # Options for negative sampling training (train.type=="negative_sampling")
 negative_sampling:
   # Type of sampler
-  sampling_type: uniform
+  sampling_type: uniform    # uniform, frequency
```
Comment (rgemulla) [inline kge/config-default.yaml]: Add a doc string that explains the sampler types.

File: kge/config-default.yaml [hunk 2]
```diff
@@ -188,7 +188,8 @@ KvsAll:
...
+  symmetric_prior: 1        # this prior is used to smooth counts in frequency sampling
```
Comment (rgemulla) [inline kge/config-default.yaml]: Rename to ``frequency.smoothing`` and add a doc string.

File: kge/util/sampler.py [hunk 3]
```diff
@@ -240,3 +241,51 @@ def _filter_and_resample_numba(
...
+    Can be smoothed with a symmetric prior.
+    """
+    def __init__(self, config, configuration_key, dataset):
+        super().__init__(config, configuration_key, dataset)
+        self.samplers = []
+        alpha = self.get_option("symmetric_prior")
+        for slot in SLOTS:
+            vocab = np.arange(self.vocabulary_size[slot].item())
+            unique, counts = np.unique(
+                dataset._triples["train"][:, slot], return_counts=True
```
Comment (rgemulla) [inline kge/util/sampler.py]: Use ``dataset.train()``

Comment (rgemulla) [reply kge/util/sampler.py]: You can get the counts directly in the format you need using ``np.bincount(y, minlength=<vocab_size>)``.

Comment (rgemulla) [reply kge/util/sampler.py]: Then simply add the smoothing value.

File: kge/util/sampler.py [hunk 4]
```diff
@@ -240,3 +241,51 @@ def _filter_and_resample_numba(
...
+            )
+
+            # smooth counts with the symmetric prior alpha
+            zero_counts = vocab[~np.isin(vocab, unique)]
+            sort_index = np.argsort(np.concatenate([unique, zero_counts]))
+            smoothed_counts = (
+                np.concatenate([counts, np.zeros(len(zero_counts))])[sort_index] + alpha
+            )
+
+            self.samplers.append(
```
Comment (rgemulla) [inline kge/util/sampler.py]: self._multinomials

**Comment by rgemulla** [comment]
Is it easy to also add the "fast" implementation? (Could also be a separate PR later on.) From my point of view, it may suffice to take this method from KgeUniformSampler and change the line starting with ``new_samples=`` accordingly.

**Comment by AdrianKs** [comment]
I addressed your comments.

I think the "fast" implementation is still some work, since in the new_samples= line we directly call the numpy random function.
For frequency sampling we would call the torch multinomial function, which probably is not supported by numba.
When we call the python sample method from the fast numba method it should become slow again.

**Comment by rgemulla** [comment]
Thanks!

**Pull request merged**